File: game/board.py
"""五子棋棋盘模块"""

import numpy as np


class Board:
    """五子棋棋盘，位置 = 行 * 宽度 + 列"""

    def __init__(self, width=15, height=15, n_in_row=5):
        self.width = width
        self.height = height
        self.n_in_row = n_in_row
        self.players = [1, 2]  # 两个玩家

    def init_board(self, start_player=0):
        """初始化棋盘"""
        self.current_player = self.players[start_player]
        self.availables = list(range(self.width * self.height))  # 所有位置可用
        self.states = {}
        self.last_move = -1

    def current_state(self):
        """获取4通道棋盘状态: [我方棋子, 对方棋子, 最后落子, 先手标记]"""
        state = np.zeros((4, self.height, self.width))

        if self.states:
            moves, players = np.array(list(zip(*self.states.items(), strict=True)))
            curr_moves = moves[players == self.current_player]
            state[0][curr_moves // self.width, curr_moves % self.width] = 1.0
            oppo_moves = moves[players != self.current_player]
            state[1][oppo_moves // self.width, oppo_moves % self.width] = 1.0
            state[2][self.last_move // self.width, self.last_move % self.width] = 1.0

        if len(self.states) % 2 == 0:
            state[3][:, :] = 1.0

        return state[:, ::-1, :]

    def do_move(self, move):
        """执行落子"""
        self.states[move] = self.current_player
        self.availables.remove(move)
        self.current_player = (
            self.players[1] if self.current_player == self.players[0] else self.players[0]
        )
        self.last_move = move

    def has_a_winner(self):
        """检查最后落子位置周围是否连成五子"""
        if self.last_move == -1:
            return False, -1

        n = self.n_in_row
        if len(self.states) < 2 * n - 1:
            return False, -1

        move = self.last_move
        h, w = move // self.width, move % self.width
        player = self.states[move]

        # 四个方向：横、竖、左斜、右斜
        directions = [(0, 1), (1, 0), (1, 1), (1, -1)]

        for dh, dw in directions:
            count = 1
            # 向一个方向延伸
            for i in range(1, n):
                nh, nw = h + i * dh, w + i * dw
                if 0 <= nh < self.height and 0 <= nw < self.width:
                    if self.states.get(nh * self.width + nw) == player:
                        count += 1
                    else:
                        break
                else:
                    break
            # 向相反方向延伸
            for i in range(1, n):
                nh, nw = h - i * dh, w - i * dw
                if 0 <= nh < self.height and 0 <= nw < self.width:
                    if self.states.get(nh * self.width + nw) == player:
                        count += 1
                    else:
                        break
                else:
                    break

            if count >= n:
                return True, player

        return False, -1

    def game_end(self):
        """检查游戏是否结束"""
        win, winner = self.has_a_winner()
        if win:
            return True, winner
        if not self.availables:
            return True, -1
        return False, -1

    def get_current_player(self):
        return self.current_player

File: src/board.js
/**
 * 五子棋棋盘逻辑 (纯 JavaScript)
 * 与 Python 端 Board 类完全一致的状态表示
 */
export class Board {
  constructor(size = 9, nInRow = 5) {
    this.size = size;
    this.nInRow = nInRow;
    this.players = [1, 2];
    this.init();
  }

  init(startPlayer = 0) {
    this.currentPlayer = this.players[startPlayer];
    this.states = new Map();
    this.availables = new Set();
    for (let i = 0; i < this.size * this.size; i++) {
      this.availables.add(i);
    }
    this.lastMove = -1;
  }

  clone() {
    const b = new Board(this.size, this.nInRow);
    b.currentPlayer = this.currentPlayer;
    b.states = new Map(this.states);
    b.availables = new Set(this.availables);
    b.lastMove = this.lastMove;
    return b;
  }

  doMove(move) {
    this.states.set(move, this.currentPlayer);
    this.availables.delete(move);
    this.currentPlayer = this.currentPlayer === this.players[0]
      ? this.players[1]
      : this.players[0];
    this.lastMove = move;
  }

  /**
   * 生成 4 通道棋盘状态 Float32Array, 与 Python 端 current_state() 一致
   * 通道0: 当前玩家棋子
   * 通道1: 对手棋子
   * 通道2: 最后落子
   * 通道3: 先手标记
   */
  currentState() {
    const s = this.size;
    const state = new Float32Array(4 * s * s);

    for (const [move, player] of this.states) {
      const r = Math.floor(move / s);
      const c = move % s;
      const flippedR = s - 1 - r; // numpy [:, ::-1, :]
      const idx = flippedR * s + c;

      if (player === this.currentPlayer) {
        state[0 * s * s + idx] = 1.0;
      } else {
        state[1 * s * s + idx] = 1.0;
      }
    }

    if (this.lastMove >= 0) {
      const r = Math.floor(this.lastMove / s);
      const c = this.lastMove % s;
      const flippedR = s - 1 - r;
      state[2 * s * s + flippedR * s + c] = 1.0;
    }

    if (this.states.size % 2 === 0) {
      for (let i = 0; i < s * s; i++) {
        state[3 * s * s + i] = 1.0;
      }
    }

    return state;
  }

  hasWinner() {
    if (this.lastMove === -1) return [false, -1];
    if (this.states.size < 2 * this.nInRow - 1) return [false, -1];

    const move = this.lastMove;
    const h = Math.floor(move / this.size);
    const w = move % this.size;
    const player = this.states.get(move);
    const n = this.nInRow;

    const directions = [[0, 1], [1, 0], [1, 1], [1, -1]];

    for (const [dh, dw] of directions) {
      let count = 1;
      for (let i = 1; i < n; i++) {
        const nh = h + i * dh, nw = w + i * dw;
        if (nh >= 0 && nh < this.size && nw >= 0 && nw < this.size &&
          this.states.get(nh * this.size + nw) === player) {
          count++;
        } else break;
      }
      for (let i = 1; i < n; i++) {
        const nh = h - i * dh, nw = w - i * dw;
        if (nh >= 0 && nh < this.size && nw >= 0 && nw < this.size &&
          this.states.get(nh * this.size + nw) === player) {
          count++;
        } else break;
      }
      if (count >= n) return [true, player];
    }
    return [false, -1];
  }

  gameEnd() {
    const [win, winner] = this.hasWinner();
    if (win) return [true, winner];
    if (this.availables.size === 0) return [true, -1];
    return [false, -1];
  }
}
